lib/classes/Utils.js

```diff
@@ -142,152 +143,167 @@ class Utils {
-              alreadyPresentEvent.count++;
-            } else {
-              numberOfEventsPerType.push({
-                name,
-                count: 1,
-              });
-            }
-          });
-
-          eventNamesPerFunction.push(funcEventsArray);
-        }
-      });
-    }
+      let userId = uuid.v1();
 
-    let hasCustomResourcesDefined = false;
-    // check if configuration in resources.Resources is defined
-    if ((serverless.service.resources &&
-      serverless.service.resources.Resources &&
-      Object.keys(serverless.service.resources.Resources).length)) {
-      hasCustomResourcesDefined = true;
-    }
-    // check if configuration in resources.Outputs is defined
-    if ((serverless.service.resources &&
-      serverless.service.resources.Outputs &&
-      Object.keys(serverless.service.resources.Outputs).length)) {
-      hasCustomResourcesDefined = true;
-    }
+      if (!this.fileExistsSync(statsEnabledFilePath)) {
+        this.writeFileSync(statsEnabledFilePath, userId);
+      } else {
+        userId = this.readFileSync(statsEnabledFilePath).toString();
+      }
 
-    let hasCustomVariableSyntaxDefined = false;
-    const defaultVariableSyntax = '\\${([ :a-zA-Z0-9._,\\-\\/\\(\\)]+?)}';
-    // check if the variableSyntax in the defaults section is defined
-    if (serverless.service.defaults &&
-      serverless.service.defaults.variableSyntax &&
-      serverless.service.defaults.variableSyntax !== defaultVariableSyntax) {
-      hasCustomVariableSyntaxDefined = true;
-    }
-    // check if the variableSyntax in the provider section is defined
-    if (serverless.service.provider &&
-      serverless.service.provider.variableSyntax &&
-      serverless.service.provider.variableSyntax !== defaultVariableSyntax) {
-      hasCustomVariableSyntaxDefined = true;
-    }
+      // function related information retrieval
+      const numberOfFunctions = _.size(serverless.service.functions);
+
+      const memorySizeAndTimeoutPerFunction = [];
+      if (numberOfFunctions) {
+        _.forEach(serverless.service.functions, (func) => {
+          const memorySize = Number(func.memorySize)
+            || Number(this.serverless.service.provider.memorySize)
+            || 1024;
+          const timeout = Number(func.timeout)
+            || Number(this.serverless.service.provider.timeout)
+            || 6;
+
+          const memorySizeAndTimeoutObject = {
+            memorySize,
+            timeout,
+          };
+
+          memorySizeAndTimeoutPerFunction.push(memorySizeAndTimeoutObject);
+        });
+      }
 
-    const auth = `${writeKey}:`;
+      // event related information retrieval
+      const numberOfEventsPerType = [];
+      const eventNamesPerFunction = [];
+      if (numberOfFunctions) {
+        _.forEach(serverless.service.functions, (func) => {
+          if (func.events) {
+            const funcEventsArray = [];
+
+            func.events.forEach((event) => {
```

Comment:
we should make sure `events` is an array, otherwise if a beginner didn't set the event as an array, he'll face a not so friendly error